### PR TITLE
8337987: Relocate jfr and throw_exception stubs from StubGenerator to SharedRuntime

### PR DESCRIPTION
**Pull request by adinn**
Store the throw_exception and jfr stub code as blobs in class SharedRuntime, move the generation code to the the arch-specific generator classes and update client code to access them from their new location.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337987](https://bugs.openjdk.org/browse/JDK-8337987): Relocate jfr and throw_exception stubs from StubGenerator to SharedRuntime (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) 🔄 Re-review required (review applies to [d0ba9688](https://git.openjdk.org/jdk/pull/20566/files/d0ba9688c736f8affb133cb9ee8b7d5afe120e15))
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Yudi Zheng](https://openjdk.org/census#yzheng) (@mur47x111 - Committer) 🔄 Re-review required (review applies to [d0ba9688](https://git.openjdk.org/jdk/pull/20566/files/d0ba9688c736f8affb133cb9ee8b7d5afe120e15))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20566/head:pull/20566` \
`$ git checkout pull/20566`

Update a local copy of the PR: \
`$ git checkout pull/20566` \
`$ git pull https://git.openjdk.org/jdk.git pull/20566/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20566`

View PR using the GUI difftool: \
`$ git pr show -t 20566`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20566.diff">https://git.openjdk.org/jdk/pull/20566.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20566#issuecomment-2286039750)

### CODE REVIEW

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp [hunk 1]
```diff
@@ -4130,11 +3914,7 @@ class StubGenerator: public StubCodeGenerator {
     StubRoutines::x86::_d2i_wrapper                = generate_d2i_wrapper(T_INT,  CAST_FROM_FN_PTR(address, SharedRuntime::d2i));
     StubRoutines::x86::_d2l_wrapper                = generate_d2i_wrapper(T_LONG, CAST_FROM_FN_PTR(address, SharedRuntime::d2l));
 
-    // Build this early so it's available for the interpreter
-    StubRoutines::_throw_StackOverflowError_entry          = generate_throw_exception("StackOverflowError throw_exception",
-                                                                                      CAST_FROM_FN_PTR(address, SharedRuntime::throw_StackOverflowError));
-    StubRoutines::_throw_delayed_StackOverflowError_entry  = generate_throw_exception("delayed StackOverflowError throw_exception",
-                                                                                      CAST_FROM_FN_PTR(address, SharedRuntime::throw_delayed_StackOverflowError));
+      CAST_FROM_FN_PTR(address, SharedRuntime::throw_delayed_StackOverflowError));
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:3917]: This does not seems correct. And I see linux-x86 (32 bits) corresponding build failure in GHA.

**Review by vnkozlov — COMMENTED** [review]
Please, add description to PR

File: src/hotspot/cpu/x86/stubGenerator_x86_32.cpp (hunk 1 above)
Comment (adinn) [reply src/hotspot/cpu/x86/stubGenerator_x86_32.cpp:3917]: Sorry, that was an accidental paste while editing the code. Should be fixed now.

**Comment by RealFYang** [comment]
Hi Andrew, I find that we need following add-on change for riscv:

```
diff --git a/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp b/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
index dc89e489b24..bed24e442e8 100644
--- a/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
+++ b/src/hotspot/cpu/riscv/sharedRuntime_riscv.cpp
@@ -66,6 +66,12 @@

 #define __ masm->

+#ifdef PRODUCT
+#define BLOCK_COMMENT(str) /* nothing */
+#else
+#define BLOCK_COMMENT(str) __ block_comment(str)
+#endif
+
 const int StackAlignmentInSlots = StackAlignmentInBytes / VMRegImpl::stack_slot_size;

 class RegisterSaver {
@@ -2742,7 +2748,7 @@ static void jfr_epilogue(MacroAssembler* masm) {
 // For c2: c_rarg0 is junk, call to runtime to write a checkpoint.
 // It returns a jobject handle to the event writer.
 // The handle is dereferenced and the return value is the event writer oop.
-static RuntimeStub* SharedRuntime::generate_jfr_write_checkpoint() {
+RuntimeStub* SharedRuntime::generate_jfr_write_checkpoint() {
   enum layout {
     fp_off,
     fp_off2,
@@ -2780,7 +2786,7 @@ static RuntimeStub* SharedRuntime::generate_jfr_write_checkpoint() {
 }

 // For c2: call to return a leased buffer.
-static RuntimeStub* SharedRuntime::generate_jfr_return_lease() {
+RuntimeStub* SharedRuntime::generate_jfr_return_lease() {
   enum layout {
     fp_off,
     fp_off2,
```

**Comment by adinn** [comment]
Hi @RealFYang

> Hi Andrew, I find that we need following add-on change for riscv:

Thanks for checking. Should be fixed by the latest push.

File: src/hotspot/share/jvmci/vmStructs_jvmci.cpp [hunk 2]
```diff
@@ -327,8 +327,6 @@
                                                                                                                                      \
   static_field(StubRoutines,                _verify_oop_count,                                jint)                                  \
                                                                                                                                      \
-  static_field(StubRoutines,                _throw_delayed_StackOverflowError_entry,          address)                               \
```
Comment (mur47x111) [inline src/hotspot/share/jvmci/vmStructs_jvmci.cpp:330]: Please add the following symbol exporting
```
diff --git a/src/hotspot/share/jvmci/jvmciCompilerToVM.hpp b/src/hotspot/share/jvmci/jvmciCompilerToVM.hpp
index 8fdb96a3038..0e96cea6596 100644
--- a/src/hotspot/share/jvmci/jvmciCompilerToVM.hpp
+++ b/src/hotspot/share/jvmci/jvmciCompilerToVM.hpp
@@ -50,6 +50,7 @@ class CompilerToVM {
     static address SharedRuntime_deopt_blob_unpack_with_exception_in_tls;
     static address SharedRuntime_deopt_blob_uncommon_trap;
     static address SharedRuntime_polling_page_return_handler;
+    static address SharedRuntime_throw_delayed_StackOverflowError_blob;
 
     static address nmethod_entry_barrier;
     static int thread_disarmed_guard_value_offset;
diff --git a/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp b/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp
index 2116133e56e..27031bf55fe 100644
--- a/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp
+++ b/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp
@@ -68,6 +68,7 @@ address CompilerToVM::Data::SharedRuntime_deopt_blob_unpack;
 address CompilerToVM::Data::SharedRuntime_deopt_blob_unpack_with_exception_in_tls;
 address CompilerToVM::Data::SharedRuntime_deopt_blob_uncommon_trap;
 address CompilerToVM::Data::SharedRuntime_polling_page_return_handler;
+address CompilerToVM::Data::SharedRuntime_throw_delayed_StackOverflowError_blob;
 
 address CompilerToVM::Data::nmethod_entry_barrier;
 int CompilerToVM::Data::thread_disarmed_guard_value_offset;
@@ -158,6 +159,7 @@ void CompilerToVM::Data::initialize(JVMCI_TRAPS) {
   SharedRuntime_deopt_blob_unpack_with_exception_in_tls = SharedRuntime::deopt_blob()->unpack_with_exception_in_tls();
   SharedRuntime_deopt_blob_uncommon_trap = SharedRuntime::deopt_blob()->uncommon_trap();
   SharedRuntime_polling_page_return_handler = SharedRuntime::polling_page_return_handler_blob()->entry_point();
+  SharedRuntime_throw_delayed_StackOverflowError_blob = SharedRuntime::throw_delayed_StackOverflowError_entry();
 
   BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
   if (bs_nm != nullptr) {
diff --git a/src/hotspot/share/jvmci/vmStructs_jvmci.cpp b/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
index 130f1032e65..e0c56ccabee 100644
--- a/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
+++ b/src/hotspot/share/jvmci/vmStructs_jvmci.cpp
@@ -68,6 +68,8 @@
   static_field(CompilerToVM::Data,             SharedRuntime_deopt_blob_uncommon_trap, address)                                      \
   static_field(CompilerToVM::Data,             SharedRuntime_polling_page_return_handler,                                            \
                                                                                        address)                                      \
+  static_field(CompilerToVM::Data,             SharedRuntime_throw_delayed_StackOverflowError_blob,                                  \
+                                                                                       address)                                      \
                                                                                                                                      \
   static_field(CompilerToVM::Data,             nmethod_entry_barrier, address)                                                       \
   static_field(CompilerToVM::Data,             thread_disarmed_guard_value_offset, int)                                              \
```

Comment (mur47x111) [reply src/hotspot/share/jvmci/vmStructs_jvmci.cpp:330]: Right, `SharedRuntime_throw_delayed_StackOverflowError_entry` is better. Thanks for the explanation!

Comment (adinn) [reply src/hotspot/share/jvmci/vmStructs_jvmci.cpp:330]: Ok, I pushed a version of your patch modified to use that name.

**Review by vnkozlov — APPROVED** [review]
Update is good.

**Review by RealFYang — APPROVED** [review]
LGTM. Thanks.

**Review by mur47x111 — APPROVED** [review]
JVMCI changes look good to me. Thanks!

**Comment by adinn** [comment]
@vnkozlov @RealFYang @mur47x111
Thank you for the reviews!

**Comment by adinn** [comment]
@vnkozlov @RealFYang Could one of you please re-review. I had to merge with master to allow for a one-line include file change in sharedRuntime.cpp.

**Review by RealFYang — APPROVED** [review]

**Comment by shqking** [comment]
Hi @adinn , I encountered Client build failure on AArch64 after this commit. Could you help take a look at it when you have spare time? Thanks.

Here shows the configuration
```
====================================================
The existing configuration has been successfully updated in
/tmp/test123/build-release
using configure arguments '--with-debug-level=release --with-version-opt=git-fe80618bf3f --with-jvm-variants=client'.

Configuration summary:
* Name:           /tmp/test123/build-release
* Debug level:    release
* HS debug level: product
* JVM variants:   client
* JVM features:   client: 'cds compiler1 epsilongc g1gc jfr jni-check jvmti management parallelgc serialgc services shenandoahgc vm-structs zgc' 
* OpenJDK target: OS: linux, CPU architecture: aarch64, address length: 64
* Version string: 24-internal-git-fe80618bf3f (24-internal)
* Source date:    1726841146 (2024-09-20T14:05:46Z)

Tools summary:
* Boot JDK:       openjdk version "22.0.2" 2024-07-16 OpenJDK Runtime Environment (build 22.0.2+9-70) OpenJDK 64-Bit Server VM (build 22.0.2+9-70, mixed mode, sharing) (at /usr/lib/jvm/jdk-22.0.2)
* Toolchain:      gcc (GNU Compiler Collection)
* C Compiler:     Version 13.2.0 (at /usr/bin/gcc)
* C++ Compiler:   Version 13.2.0 (at /usr/bin/g++)

Build performance summary:
* Build jobs:     72
* Memory limit:   587068 MB
```

And here shows the error msg:
```
=== Output from failing command(s) repeated here ===
* For target hotspot_variant-client_libjvm_objs_sharedRuntime_aarch64.o:
/tmp/test123/jdk-src/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp: In static member function ‘static RuntimeStub* SharedRuntime::generate_throw_exception(const char*, address)’:
/tmp/test123/jdk-src/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:2809:3: error: ‘TraceTime’ was not declared in this scope; did you mean ‘traceid’?
 2809 |   TraceTime timer(timer_msg, TRACETIME_LOG(Info, startuptime));
      |   ^~~~~~~~~
      |   traceid

* All command lines available in /tmp/test123/build-release/make-support/failure-logs.
=== End of repeated output ===
```

**Comment by adinn** [comment]
@shqking Thanks for reporting this problem. I reproduced the same failure on aarch64 and have noticed that the problem also arises on arm. File `sharedRuntime_aarch64.cpp` needs to include the header that declares class TraceTime as does `sharedRuntime_arm.cpp`.

I raised JDK-8340793 and will push a patch to fix it.

**Pull request closed without merging**